File: sdk/consensus/circuit_breaker.py
"""
Circuit Breaker Pattern for AI/ML Layer Operations.

Matches luxtensor-consensus/src/circuit_breaker.rs.

Prevents cascade failures when AI/ML operations timeout or fail.
Implements the Circuit Breaker pattern with three states:
- Closed: Normal operation, requests pass through
- Open: Too many failures, requests fail fast
- HalfOpen: Testing if service recovered
"""

from dataclasses import dataclass
from enum import Enum
from typing import Callable, Optional, TypeVar
import time
import threading

T = TypeVar('T')
E = TypeVar('E', bound=Exception)


class CircuitState(str, Enum):
    """Circuit breaker states."""
    CLOSED = "closed"       # Normal operation - requests pass through
    OPEN = "open"           # Too many failures - requests fail fast
    HALF_OPEN = "half_open"  # Testing if service recovered


@dataclass
class CircuitBreakerConfig:
    """
    Configuration for circuit breaker.

    Matches luxtensor-consensus CircuitBreakerConfig.
    """
    name: str = "default"
    failure_threshold: int = 5          # Failures before opening
    success_threshold: int = 3          # Successes in half-open to close
    open_duration_secs: float = 30.0    # Seconds to stay open
    timeout_secs: float = 10.0          # Operation timeout

    @classmethod
    def default(cls) -> 'CircuitBreakerConfig':
        return cls()

    def with_failure_threshold(self, threshold: int) -> 'CircuitBreakerConfig':
        self.failure_threshold = threshold
        return self

    def with_open_duration(self, secs: float) -> 'CircuitBreakerConfig':
        self.open_duration_secs = secs
        return self

    def with_timeout(self, secs: float) -> 'CircuitBreakerConfig':
        self.timeout_secs = secs
        return self

    def with_success_threshold(self, threshold: int) -> 'CircuitBreakerConfig':
        self.success_threshold = threshold
        return self


@dataclass
class CircuitBreakerStats:
    """
    Statistics for monitoring.

    Matches luxtensor-consensus CircuitBreakerStats.
    """
    total_requests: int = 0
    successful_requests: int = 0
    failed_requests: int = 0
    rejected_requests: int = 0
    current_failures: int = 0
    current_successes: int = 0
    state_changes: int = 0
    last_failure_time: Optional[float] = None
    last_state_change_time: Optional[float] = None

    def to_dict(self) -> dict:
        return {
            "total_requests": self.total_requests,
            "successful_requests": self.successful_requests,
            "failed_requests": self.failed_requests,
            "rejected_requests": self.rejected_requests,
            "success_rate": (
                self.successful_requests / self.total_requests * 100
                if self.total_requests > 0 else 0
            ),
            "current_failures": self.current_failures,
            "current_successes": self.current_successes,
            "state_changes": self.state_changes,
            "last_failure_time": self.last_failure_time,
        }


class CircuitBreakerError(Exception):
    """Errors from circuit breaker."""
    pass


class CircuitOpenError(CircuitBreakerError):
    """Circuit is open, request rejected."""
    pass


class OperationTimeoutError(CircuitBreakerError):
    """Operation timed out."""
    pass


class CircuitBreaker:
    """
    Circuit Breaker implementation.

    Matches luxtensor-consensus CircuitBreaker.

    Usage:
        cb = CircuitBreaker.with_name("ai_inference")

        try:
            result = cb.execute(lambda: ai_model.inference(data))
        except CircuitOpenError:
            # Use fallback
            result = fallback_value
    """

    def __init__(self, config: Optional[CircuitBreakerConfig] = None):
        self.config = config or CircuitBreakerConfig.default()
        self._state = CircuitState.CLOSED
        self._stats = CircuitBreakerStats()
        self._opened_at: Optional[float] = None
        self._lock = threading.RLock()

    @classmethod
    def with_name(cls, name: str) -> 'CircuitBreaker':
        """Create with named config."""
        config = CircuitBreakerConfig(name=name)
        return cls(config)

    @property
    def state(self) -> CircuitState:
        """Get current state."""
        with self._lock:
            self._check_and_update_state()
            return self._state

    def allow_request(self) -> bool:
        """Check if request should be allowed."""
        with self._lock:
            self._check_and_update_state()

            if self._state == CircuitState.CLOSED:
                return True
            elif self._state == CircuitState.HALF_OPEN:
                return True  # Allow test request
            else:  # OPEN
                return False

    def record_success(self):
        """Record a successful operation."""
        with self._lock:
            self._stats.total_requests += 1
            self._stats.successful_requests += 1
            self._stats.current_failures = 0
            self._stats.current_successes += 1

            if self._state == CircuitState.HALF_OPEN:
                if self._stats.current_successes >= self.config.success_threshold:
                    self._transition_to(CircuitState.CLOSED)

    def record_failure(self):
        """Record a failed operation."""
        with self._lock:
            self._stats.total_requests += 1
            self._stats.failed_requests += 1
            self._stats.current_failures += 1
            self._stats.current_successes = 0
            self._stats.last_failure_time = time.time()

            if self._state == CircuitState.HALF_OPEN:
                # Immediate open on failure in half-open
                self._transition_to(CircuitState.OPEN)
            elif self._state == CircuitState.CLOSED:
                if self._stats.current_failures >= self.config.failure_threshold:
                    self._transition_to(CircuitState.OPEN)

    def record_rejected(self):
        """Record a rejected request (circuit open)."""
        with self._lock:
            self._stats.total_requests += 1
            self._stats.rejected_requests += 1

    def execute(self, operation: Callable[[], T]) -> T:
        """
        Execute operation with circuit breaker protection.

        Args:
            operation: Function to execute

        Returns:
            Result of operation

        Raises:
            CircuitOpenError: If circuit is open
            Exception: If operation fails
        """
        if not self.allow_request():
            self.record_rejected()
            raise CircuitOpenError(f"Circuit '{self.config.name}' is open")

        try:
            result = operation()
            self.record_success()
            return result
        except Exception:
            self.record_failure()
            raise

    def execute_with_fallback(
        self,
        operation: Callable[[], T],
        fallback: T,
    ) -> T:
        """Execute with fallback value when circuit is open."""
        try:
            return self.execute(operation)
        except CircuitOpenError:
            return fallback
        except Exception:
            return fallback

    @property
    def stats(self) -> CircuitBreakerStats:
        """Get statistics."""
        with self._lock:
            return CircuitBreakerStats(
                total_requests=self._stats.total_requests,
                successful_requests=self._stats.successful_requests,
                failed_requests=self._stats.failed_requests,
                rejected_requests=self._stats.rejected_requests,
                current_failures=self._stats.current_failures,
                current_successes=self._stats.current_successes,
                state_changes=self._stats.state_changes,
                last_failure_time=self._stats.last_failure_time,
                last_state_change_time=self._stats.last_state_change_time,
            )

    def reset(self):
        """Reset circuit breaker."""
        with self._lock:
            self._state = CircuitState.CLOSED
            self._stats = CircuitBreakerStats()
            self._opened_at = None

    def _check_and_update_state(self):
        """Check if circuit should transition from Open to HalfOpen."""
        if self._state == CircuitState.OPEN and self._opened_at:
            elapsed = time.time() - self._opened_at
            if elapsed >= self.config.open_duration_secs:
                self._transition_to(CircuitState.HALF_OPEN)

    def _transition_to(self, new_state: CircuitState):
        """Transition to a new state."""
        if self._state == new_state:
            return

        _ = self._state  # Capture current state before transition
        self._state = new_state
        self._stats.state_changes += 1
        self._stats.last_state_change_time = time.time()

        if new_state == CircuitState.OPEN:
            self._opened_at = time.time()
        elif new_state == CircuitState.CLOSED:
            self._stats.current_failures = 0
            self._stats.current_successes = 0
        elif new_state == CircuitState.HALF_OPEN:
            self._stats.current_successes = 0


class CircuitBreakerRegistry:
    """
    Registry for managing multiple circuit breakers.
    """

    _instance: Optional['CircuitBreakerRegistry'] = None
    _lock = threading.Lock()

    def __new__(cls):
        with cls._lock:
            if cls._instance is None:
                cls._instance = super().__new__(cls)
                cls._instance._breakers = {}
                cls._instance._registry_lock = threading.RLock()
            return cls._instance

    def get(self, name: str) -> CircuitBreaker:
        """Get or create a circuit breaker by name."""
        with self._registry_lock:
            if name not in self._breakers:
                self._breakers[name] = CircuitBreaker.with_name(name)
            return self._breakers[name]

    def get_all_stats(self) -> dict:
        """Get stats for all circuit breakers."""
        with self._registry_lock:
            return {
                name: cb.stats.to_dict()
                for name, cb in self._breakers.items()
            }

    def reset_all(self):
        """Reset all circuit breakers."""
        with self._registry_lock:
            for cb in self._breakers.values():
                cb.reset()

    def remove(self, name: str) -> bool:
        """
        Remove a circuit breaker from the registry.

        Args:
            name: Name of the circuit breaker to remove

        Returns:
            True if removed, False if not found
        """
        with self._registry_lock:
            if name in self._breakers:
                del self._breakers[name]
                return True
            return False

    def clear(self) -> int:
        """
        Remove all circuit breakers from the registry.

        Returns:
            Number of circuit breakers removed
        """
        with self._registry_lock:
            count = len(self._breakers)
            self._breakers.clear()
            return count


# Global registry
def get_circuit_breaker(name: str) -> CircuitBreaker:
    """Get a circuit breaker by name from the global registry."""
    return CircuitBreakerRegistry().get(name)


# Module exports
__all__ = [
    "CircuitState",
    "CircuitBreakerConfig",
    "CircuitBreakerStats",
    "CircuitBreakerError",
    "CircuitOpenError",
    "OperationTimeoutError",
    "CircuitBreaker",
    "CircuitBreakerRegistry",
    "get_circuit_breaker",
]

File: luxtensor/crates/luxtensor-consensus/src/circuit_breaker.rs
//! Circuit Breaker Pattern for AI/ML Layer Operations
//!
//! Prevents cascade failures when AI/ML operations timeout or fail.
//! Implements the Circuit Breaker pattern with three states:
//! - Closed: Normal operation, requests pass through
//! - Open: Failures exceeded threshold, requests fail fast
//! - HalfOpen: Testing if service recovered
//!
//! Reference: Architecture patterns for distributed systems resilience

use std::sync::atomic::{AtomicU32, AtomicU64, Ordering};
use std::time::{Duration, Instant};
use parking_lot::RwLock;
use tracing::{info, warn};

/// Circuit breaker states
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CircuitState {
    /// Normal operation - requests pass through
    Closed,
    /// Too many failures - requests fail fast
    Open,
    /// Testing if service recovered
    HalfOpen,
}

impl Default for CircuitState {
    fn default() -> Self {
        Self::Closed
    }
}

/// Configuration for circuit breaker
#[derive(Debug, Clone)]
pub struct CircuitBreakerConfig {
    /// Number of failures before opening circuit
    pub failure_threshold: u32,
    /// Duration to keep circuit open before testing
    pub open_duration: Duration,
    /// Number of successful requests to close circuit
    pub success_threshold: u32,
    /// Timeout for individual operations
    pub operation_timeout: Duration,
    /// Name for logging
    pub name: String,
}

impl Default for CircuitBreakerConfig {
    fn default() -> Self {
        Self {
            failure_threshold: 5,
            open_duration: Duration::from_secs(30),
            success_threshold: 3,
            operation_timeout: Duration::from_secs(10),
            name: "default".to_string(),
        }
    }
}

impl CircuitBreakerConfig {
    pub fn new(name: &str) -> Self {
        Self {
            name: name.to_string(),
            ..Default::default()
        }
    }

    pub fn with_failure_threshold(mut self, threshold: u32) -> Self {
        self.failure_threshold = threshold;
        self
    }

    pub fn with_open_duration(mut self, duration: Duration) -> Self {
        self.open_duration = duration;
        self
    }

    pub fn with_timeout(mut self, timeout: Duration) -> Self {
        self.operation_timeout = timeout;
        self
    }

    pub fn with_success_threshold(mut self, threshold: u32) -> Self {
        self.success_threshold = threshold;
        self
    }
}

/// Statistics for monitoring
#[derive(Debug, Clone, Default)]
pub struct CircuitBreakerStats {
    pub total_requests: u64,
    pub successful_requests: u64,
    pub failed_requests: u64,
    pub rejected_requests: u64,
    pub state_changes: u32,
    pub current_state: CircuitState,
}

/// Circuit Breaker implementation
///
/// NOTE: Uses `Instant::now()` intentionally — circuit breaker timings are local
/// to each node and do not affect consensus determinism. The open-to-half-open
/// transition is based on wall-clock elapsed time, which is acceptable because
/// it only governs local request gating, not any on-chain state transition.
pub struct CircuitBreaker {
    config: CircuitBreakerConfig,
    /// Current state
    state: RwLock<CircuitState>,
    /// Consecutive failure count
    failure_count: AtomicU32,
    /// Consecutive success count (in half-open state)
    success_count: AtomicU32,
    /// Time when circuit was opened
    opened_at: RwLock<Option<Instant>>,
    /// Statistics
    total_requests: AtomicU64,
    successful_requests: AtomicU64,
    failed_requests: AtomicU64,
    rejected_requests: AtomicU64,
    state_changes: AtomicU32,
}

impl CircuitBreaker {
    /// Create a new circuit breaker
    pub fn new(config: CircuitBreakerConfig) -> Self {
        info!("CircuitBreaker '{}' initialized", config.name);
        Self {
            config,
            state: RwLock::new(CircuitState::Closed),
            failure_count: AtomicU32::new(0),
            success_count: AtomicU32::new(0),
            opened_at: RwLock::new(None),
            total_requests: AtomicU64::new(0),
            successful_requests: AtomicU64::new(0),
            failed_requests: AtomicU64::new(0),
            rejected_requests: AtomicU64::new(0),
            state_changes: AtomicU32::new(0),
        }
    }

    /// Create with default config
    pub fn with_name(name: &str) -> Self {
        Self::new(CircuitBreakerConfig::new(name))
    }

    /// Get current state
    pub fn state(&self) -> CircuitState {
        self.check_and_update_state();
        *self.state.read()
    }

    /// Check if request should be allowed
    pub fn allow_request(&self) -> bool {
        self.check_and_update_state();

        let state = *self.state.read();
        match state {
            CircuitState::Closed => true,
            CircuitState::Open => false,
            CircuitState::HalfOpen => true, // Allow test requests
        }
    }

    /// Record a successful operation
    ///
    /// SECURITY: Holds state write lock across the entire operation to prevent
    /// TOCTOU races between state read and counter updates.
    pub fn record_success(&self) {
        self.total_requests.fetch_add(1, Ordering::Relaxed);
        self.successful_requests.fetch_add(1, Ordering::Relaxed);

        // Hold write lock to ensure atomic read-modify-write of state + counters
        let state = self.state.write();
        self.failure_count.store(0, Ordering::SeqCst);

        if *state == CircuitState::HalfOpen {
            let successes = self.success_count.fetch_add(1, Ordering::SeqCst) + 1;
            if successes >= self.config.success_threshold {
                drop(state); // drop before transition_to which also takes write lock
                self.transition_to(CircuitState::Closed);
                return;
            }
        }
        drop(state);
    }

    /// Record a failed operation
    ///
    /// SECURITY: Holds state write lock across the entire operation to prevent
    /// TOCTOU races between state read and counter updates.
    pub fn record_failure(&self) {
        self.total_requests.fetch_add(1, Ordering::Relaxed);
        self.failed_requests.fetch_add(1, Ordering::Relaxed);

        // Hold write lock to ensure atomic read-modify-write of state + counters
        let state = self.state.write();
        self.success_count.store(0, Ordering::SeqCst);

        match *state {
            CircuitState::Closed => {
                let failures = self.failure_count.fetch_add(1, Ordering::SeqCst) + 1;
                if failures >= self.config.failure_threshold {
                    drop(state);
                    self.transition_to(CircuitState::Open);
                    return;
                }
            }
            CircuitState::HalfOpen => {
                // Any failure in half-open reopens circuit
                drop(state);
                self.transition_to(CircuitState::Open);
                return;
            }
            CircuitState::Open => {
                // Already open, just count
            }
        }
        drop(state);
    }

    /// Record a rejected request (circuit open)
    pub fn record_rejected(&self) {
        self.total_requests.fetch_add(1, Ordering::Relaxed);
        self.rejected_requests.fetch_add(1, Ordering::Relaxed);
    }

    /// Execute operation with circuit breaker protection
    pub fn execute<F, T, E>(&self, operation: F) -> Result<T, CircuitBreakerError<E>>
    where
        F: FnOnce() -> Result<T, E>,
    {
        if !self.allow_request() {
            self.record_rejected();
            return Err(CircuitBreakerError::CircuitOpen);
        }

        match operation() {
            Ok(result) => {
                self.record_success();
                Ok(result)
            }
            Err(e) => {
                self.record_failure();
                Err(CircuitBreakerError::OperationFailed(e))
            }
        }
    }

    /// Execute with fallback value when circuit is open
    pub fn execute_with_fallback<F, T, E>(&self, operation: F, fallback: T) -> T
    where
        F: FnOnce() -> Result<T, E>,
        T: Clone,
    {
        match self.execute(operation) {
            Ok(result) => result,
            Err(_) => fallback,
        }
    }

    /// Get statistics
    pub fn stats(&self) -> CircuitBreakerStats {
        CircuitBreakerStats {
            total_requests: self.total_requests.load(Ordering::SeqCst),
            successful_requests: self.successful_requests.load(Ordering::SeqCst),
            failed_requests: self.failed_requests.load(Ordering::SeqCst),
            rejected_requests: self.rejected_requests.load(Ordering::SeqCst),
            state_changes: self.state_changes.load(Ordering::SeqCst),
            current_state: self.state(),
        }
    }

    /// Reset circuit breaker
    pub fn reset(&self) {
        *self.state.write() = CircuitState::Closed;
        *self.opened_at.write() = None;
        self.failure_count.store(0, Ordering::SeqCst);
        self.success_count.store(0, Ordering::SeqCst);
        info!("CircuitBreaker '{}' reset", self.config.name);
    }

    /// Check if circuit should transition from Open to HalfOpen
    fn check_and_update_state(&self) {
        let state = *self.state.read();
        if state == CircuitState::Open {
            if let Some(opened_at) = *self.opened_at.read() {
                if opened_at.elapsed() >= self.config.open_duration {
                    self.transition_to(CircuitState::HalfOpen);
                }
            }
        }
    }

    /// Transition to a new state
    fn transition_to(&self, new_state: CircuitState) {
        let old_state = *self.state.read();
        if old_state == new_state {
            return;
        }

        *self.state.write() = new_state;
        self.state_changes.fetch_add(1, Ordering::SeqCst);

        match new_state {
            CircuitState::Open => {
                *self.opened_at.write() = Some(Instant::now());
                self.success_count.store(0, Ordering::SeqCst);
                warn!(
                    "CircuitBreaker '{}': {} -> Open (failures: {})",
                    self.config.name,
                    format!("{:?}", old_state),
                    self.failure_count.load(Ordering::SeqCst)
                );
            }
            CircuitState::HalfOpen => {
                self.success_count.store(0, Ordering::SeqCst);
                info!(
                    "CircuitBreaker '{}': Open -> HalfOpen (testing recovery)",
                    self.config.name
                );
            }
            CircuitState::Closed => {
                *self.opened_at.write() = None;
                self.failure_count.store(0, Ordering::SeqCst);
                info!(
                    "CircuitBreaker '{}': {} -> Closed (recovered)",
                    self.config.name,
                    format!("{:?}", old_state)
                );
            }
        }
    }
}

/// Errors from circuit breaker
#[derive(Debug)]
pub enum CircuitBreakerError<E> {
    /// Circuit is open, request rejected
    CircuitOpen,
    /// Operation failed with inner error
    OperationFailed(E),
    /// Operation timed out
    Timeout,
}

impl<E: std::fmt::Display> std::fmt::Display for CircuitBreakerError<E> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::CircuitOpen => write!(f, "Circuit breaker is open"),
            Self::OperationFailed(e) => write!(f, "Operation failed: {}", e),
            Self::Timeout => write!(f, "Operation timed out"),
        }
    }
}

impl<E: std::fmt::Debug + std::fmt::Display> std::error::Error for CircuitBreakerError<E> {}

/// AI Layer Circuit Breaker - specialized for AI/ML operations
pub struct AILayerCircuitBreaker {
    /// Circuit breaker for weight consensus
    pub weight_consensus: CircuitBreaker,
    /// Circuit breaker for commit-reveal
    pub commit_reveal: CircuitBreaker,
    /// Circuit breaker for emission distribution
    pub emission: CircuitBreaker,
}

impl AILayerCircuitBreaker {
    pub fn new() -> Self {
        Self {
            weight_consensus: CircuitBreaker::new(
                CircuitBreakerConfig::new("weight_consensus")
                    .with_failure_threshold(3)
                    .with_open_duration(Duration::from_secs(60))
                    .with_timeout(Duration::from_secs(30))
            ),
            commit_reveal: CircuitBreaker::new(
                CircuitBreakerConfig::new("commit_reveal")
                    .with_failure_threshold(5)
                    .with_open_duration(Duration::from_secs(120))
                    .with_timeout(Duration::from_secs(60))
            ),
            emission: CircuitBreaker::new(
                CircuitBreakerConfig::new("emission")
                    .with_failure_threshold(3)
                    .with_open_duration(Duration::from_secs(30))
                    .with_timeout(Duration::from_secs(15))
            ),
        }
    }

    /// Get overall AI layer health
    pub fn is_healthy(&self) -> bool {
        self.weight_consensus.state() == CircuitState::Closed
            && self.commit_reveal.state() == CircuitState::Closed
            && self.emission.state() == CircuitState::Closed
    }

    /// Get summary of all circuit breakers
    pub fn summary(&self) -> AILayerStatus {
        AILayerStatus {
            weight_consensus_state: self.weight_consensus.state(),
            commit_reveal_state: self.commit_reveal.state(),
            emission_state: self.emission.state(),
            healthy: self.is_healthy(),
        }
    }
}

impl Default for AILayerCircuitBreaker {
    fn default() -> Self {
        Self::new()
    }
}

/// Status of AI layer circuit breakers
#[derive(Debug, Clone)]
pub struct AILayerStatus {
    pub weight_consensus_state: CircuitState,
    pub commit_reveal_state: CircuitState,
    pub emission_state: CircuitState,
    pub healthy: bool,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_circuit_breaker_starts_closed() {
        let cb = CircuitBreaker::with_name("test");
        assert_eq!(cb.state(), CircuitState::Closed);
        assert!(cb.allow_request());
    }

    #[test]
    fn test_circuit_opens_after_failures() {
        let cb = CircuitBreaker::new(
            CircuitBreakerConfig::new("test")
                .with_failure_threshold(3)
        );

        // Record failures
        cb.record_failure();
        cb.record_failure();
        assert_eq!(cb.state(), CircuitState::Closed);

        cb.record_failure(); // 3rd failure
        assert_eq!(cb.state(), CircuitState::Open);
        assert!(!cb.allow_request());
    }

    #[test]
    fn test_circuit_transitions_to_half_open() {
        let cb = CircuitBreaker::new(
            CircuitBreakerConfig::new("test")
                .with_failure_threshold(1)
                .with_open_duration(Duration::from_millis(50))
        );

        cb.record_failure();
        assert_eq!(cb.state(), CircuitState::Open);

        // Wait for open duration
        std::thread::sleep(Duration::from_millis(60));

        // Should transition to half-open
        assert_eq!(cb.state(), CircuitState::HalfOpen);
        assert!(cb.allow_request());
    }

    #[test]
    fn test_circuit_closes_after_successes() {
        let cb = CircuitBreaker::new(
            CircuitBreakerConfig::new("test")
                .with_failure_threshold(1)
                .with_success_threshold(2)
                .with_open_duration(Duration::from_millis(10))
        );

        // Open the circuit
        cb.record_failure();
        assert_eq!(cb.state(), CircuitState::Open);

        // Wait and transition to half-open
        std::thread::sleep(Duration::from_millis(20));
        assert_eq!(cb.state(), CircuitState::HalfOpen);

        // Record successes
        cb.record_success();
        assert_eq!(cb.state(), CircuitState::HalfOpen);

        cb.record_success(); // 2nd success
        assert_eq!(cb.state(), CircuitState::Closed);
    }

    #[test]
    fn test_half_open_failure_reopens() {
        let cb = CircuitBreaker::new(
            CircuitBreakerConfig::new("test")
                .with_failure_threshold(1)
                .with_open_duration(Duration::from_millis(10))
        );

        cb.record_failure();
        std::thread::sleep(Duration::from_millis(20));
        assert_eq!(cb.state(), CircuitState::HalfOpen);

        // Failure in half-open reopens
        cb.record_failure();
        assert_eq!(cb.state(), CircuitState::Open);
    }

    #[test]
    fn test_execute_with_fallback() {
        let cb = CircuitBreaker::new(
            CircuitBreakerConfig::new("test")
                .with_failure_threshold(1)
        );

        // Open the circuit
        cb.record_failure();

        // Execute should return fallback
        let result = cb.execute_with_fallback(|| Ok::<i32, ()>(42), 0);
        assert_eq!(result, 0);
    }

    #[test]
    fn test_statistics() {
        let cb = CircuitBreaker::with_name("test");

        cb.record_success();
        cb.record_success();
        cb.record_failure();

        let stats = cb.stats();
        assert_eq!(stats.total_requests, 3);
        assert_eq!(stats.successful_requests, 2);
        assert_eq!(stats.failed_requests, 1);
    }

    #[test]
    fn test_ai_layer_circuit_breaker() {
        let ai_cb = AILayerCircuitBreaker::new();

        assert!(ai_cb.is_healthy());

        // Trip weight consensus
        for _ in 0..3 {
            ai_cb.weight_consensus.record_failure();
        }

        assert!(!ai_cb.is_healthy());

        let status = ai_cb.summary();
        assert_eq!(status.weight_consensus_state, CircuitState::Open);
        assert_eq!(status.commit_reveal_state, CircuitState::Closed);
    }

    #[test]
    fn test_reset() {
        let cb = CircuitBreaker::new(
            CircuitBreakerConfig::new("test")
                .with_failure_threshold(1)
        );

        cb.record_failure();
        assert_eq!(cb.state(), CircuitState::Open);

        cb.reset();
        assert_eq!(cb.state(), CircuitState::Closed);
        assert!(cb.allow_request());
    }
}
